src/v/cluster/scheduling/constraints.cc

```diff
@@ -29,8 +29,10 @@ hard_constraint not_fully_allocated() {
     class impl : public hard_constraint::impl {
     public:
         hard_constraint_evaluator
-        make_evaluator(const replicas_t&) const final {
-            return [](const allocation_node& node) { return !node.is_full(); };
+        make_evaluator(const model::ntp& ntp, const replicas_t&) const final {
+            return [&ntp](const allocation_node& node) {
+                return !node.is_full(ntp);
```

Comment:
nit: not necessary to fix right now, but I have a feeling that `allocation_state` should be just "dumb state" (e.g. for full node check it should just provide the number of currently allocated partitions) and all logic should be in evaluators. The reason is that 1) evaluators already contain some of the logic, why not all? and 2) they have more context available (existing replicas, previous node for that replica (not available right now, but we'll have to add it soon)) so we'll either have to pass it all to `allocation_state` or just decide everything in the evaluators.